### ISSUE TITLE
[@types/react] Prameter type definition  between useMemo and useCallback is different, but the origin is same

### ISSUE THREAD

**Issue by khg0712** [DefinitelyTyped/DefinitelyTyped#45283]
Is something wrong with CI or publishing infrastructure?

- Check the [Infrastructure status updates](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/44317) thread for known issues.
- Leave a message on the [DefinitelyTyped Gitter](http://gitter.im/DefinitelyTyped/DefinitelyTyped).

If you know how to fix the issue, make a pull request instead.

- [x] I tried using the `@types/react` package and had problems.
- [ ] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [ ] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @johnnyreilly @bbenezech @pzavolinsky @digiguru @ericanderson @DovydasNavickas @theruther4d @guilhermehubner @ferdaber @jrakotoharisoa @pascaloliv @hotell @franklixuefei @Jessidhia @saranshkataria @lukyth @eps1lon @zieka @dancerphil 

---
## updated issue

I recognized why it should have different type each @types/react and react. (It means not difference of flow and typescript but difference of type definition between @types/react and react.)

My issue is that param type of useCallback and useMemo is same in react. but, it is different in @types/react.
I think it is inconsistent. because the reason of difference between @types/react and react only applies to useCallback. I think param type of useMemo should be changed.

---
## origin issue

I have watched origin code of react hooks (specipically useCallback) [here](https://github.com/facebook/react/blob/4c7036e807fa18a3e21a5182983c7c0f05c5936e/packages/react/src/ReactHooks.js#L120-L126).
But the type of `useCallback` was different from [useCallback](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/cdd1ff9d70c4dd9327af83e0940e697b2704e374/types/react/index.d.ts#L1104) of `@types/react`

```ts
type DependencyList = ReadonlyArray<any>;
function useCallback<T extends (...args: any[]) => any>(
  callback: T,
  deps: DependencyList
): T;
```
type definition of @types/react

```js
function useCallback<T>(
  callback: T,
  deps: Array<mixed> | void | null, //void is undefined in flow
): T
```
type definition of react(origin)

Why `useCallback` has different type definition from origin? and other hooks are different too.

ps. If it should be fixed, I want to fix it. thx

**Comment by eps1lon** [comment]
> Why useCallback has different type definition from origin?

The React repo is typed in `flow`. `flow` has slightly different syntax which explains the literal difference in code.

Is there a specific type problem you're facing?

**Comment by ferdaber** [comment]
sure in real life you could call `useCallback` or `useMemo` without a dependency array, but in practice you would never want to do that, since at that point those are just IIFEs.

**Comment by khg0712** [comment]
@ferdaber 
Now, It's clear to me. I thought that type definition of `@types/react` should implement type definition of origin `react`. 

But now I got understanding difference of type definition is for following intention of react hooks. 👍

**Comment by khg0712** [comment]
@ferdaber 
Then both `useCallback` and `useMemo` should have *notnull* deps param.
But, deps param of  `useMemo`  is `DependencyList | undefined` (see [here](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/58cd2a2af7441ff51f37789e21acdd41731c13c4/types/react/index.d.ts#L1124))
I think it is very contradictory.

**Comment by eps1lon** [comment]
> ps. Actually, there is a comment deps param of useMemo allows undefined, but it is mostly a mistake.

Yeah, I'm not happy with that comment either. It doesn't explain what the actual use case would be. Just that there is one without considering it might not actually be one. 

Maybe @Jessidhia can shed some light?

**Comment by ferdaber** [comment]
IIRC someone reported a 0.01% where they wanted to explicitly pass in `undefined`, and they wanted to be able to do it without using the non null assertion parameter, so this was the compromise that we had: forcing the second parameter, but allowing `undefined`.

**Comment by khg0712** [comment]
> so this was the compromise that we had: forcing the second parameter, but allowing `undefined`.

@ferdaber 

Such rules must be consistent. `useCallback` isn't allowed `undefined` but `useMemo` allowed `undefined`. Same rules must be applied to not only `useMemo`, but also `useCallback`.

**Comment by eps1lon** [comment]
> Such rules must be consistent. 

There's no "must" here. React core members even approved the original change so I don't think we need to change something because you're of the opinion that we "must" do something.

**Comment by khg0712** [comment]
> There's no "must" here. React core members even approved the original change so I don't think we need to change something because you're of the opinion that we "must" do something.

@eps1lon I'm sorry if you felt forced.It wasn't intended. I thought it was weird that different rules are applied to same condition.

But if you recognized it and it is no matter to you. It is ok. It would be ok to close this issue.🙂

**Comment by gapurov** [comment]
I would request @khg0712 to reopen this issue and humbly ask @eps1lon to reconsider his position. 

I think to dismiss the arguments of @khg0712 as non-issue because react core members approved the original change is not right. 

Types are written to represent an API not to save users from potential misusing the functionality or according to how many people use a feature. For example the [core code](https://github.com/facebook/react/blob/e12a9dfc96be12ea8e5c759986041ee5308e8e06/packages/react-reconciler/src/ReactFiberHooks.new.js#L1874) of the `useCallback`, has an explicit check: `const nextDeps = deps === undefined ? null : deps;`, the same check has useMemo. This is reflected in the flow types. Why should DT say otherwise and disallow the use intended by creators?

**Comment by Hideman42** [comment]
I'm also having issue with the typing of the useCallback.

The older [function useCallback<T extends (...args: any[]) => any>(callback: T, deps: DependencyList): T;](https://github.com/DefinitelyTyped/DefinitelyTyped/blame/45cb0a9afa0130c7f4b13c0370024009b5f29e1f/types/react/index.d.ts#L1113C5-L1113C99)

Would be preferred over [function useCallback<T extends Function>(callback: T, deps: DependencyList): T;](https://github.com/DefinitelyTyped/DefinitelyTyped/blame/master/types/react/index.d.ts#L1813)

```ts
type DependencyList = any[];

function useCallback<T extends Function>(callback: T, deps: DependencyList): T { return callback; }
function useCallbackOld<T extends (...args: any[]) => any>(callback: T, deps: DependencyList): T { return callback; }

function test(func: ((arg: string) => string) | undefined): string | undefined { return func?.('test'); }

const a = () => test(useCallback(str => str.toUpperCase(), [])) // str implicit type any
const b = () => test(useCallbackOld(str => str.toUpperCase(), []))
```
[Here a TS playground](https://www.typescriptlang.org/play/?ts=5.9.3#code/C4TwDgpgBAIhkDsAmEEGMQBkCWBnYUAvFAIYIgDaAugNwBQdAZgK7rDYD2CUzuEAwiQA2QgEYk0AawA8AFSgQAHsFRJcUAGKs07LgD4AFGmFiJkgFxRZAGigowuS3EQp0WPMACUl+QG8oAE4QwMwB3MYi4lI0UAC+TNq63LwCJlGSAPJCSHIKyqrqBgB0JSQBAOaOpOTUnkR61SCGEaZSPrb2Vc6qqBg4+N5WUP5BIWFQLekx8QlsnNwq+AYs6JYGBmXllvgB2AjldYQNO3sHUAA+PMgQjHsQSIMn+xdXKLcI98OBwaHcK2gAfiKBgA5ItgCDPNMGGguPhSEQoAZDg1wQYUoJImYDDt6lAdkVgBwAKpgSABQR8ZG2Wp1AD0dPxwACUGwAFswEJsGhsARQJBGnRYQh4aJEci8WiMWkzFkkDjmXiCUTSeTKRBqVBaXQgA)
